modules/core/test/com/haulmont/cuba/core/MessagesTest.java

```diff
@@ -178,7 +199,7 @@ public void testCachingDefaultLocSeveralPacks() {
 
         String msg = messages.getMessage("com.haulmont.cuba.core.mp_test.nested com.haulmont.cuba.core.mp_test", "key0");
         assertEquals("Message0", msg);
-        assertEquals(14, getSearchMessagesCount());
+        assertEquals(18, getSearchMessagesCount());
```

Comment:
Your fix adds an extra run to search for a message, it is not good.
The problem is that the default language is Russian in your application and English in the CUBA application. And when you use the `@include` in the default message pack for your application (RU), then the search will take place in the default message pack of CUBA application (EN).
I agree with that this is a bug, but it can be solved differently so that the number of runs to search for a message does not increase.
Add the following changes:
  * `AbstractMessages#searchFiles`
    Replace `Properties properties = loadPropertiesFromFile(packPath, locale, truncatedLocale);` line with:
      ```
      Properties properties = loadPropertiesFromFile(packPath, locale, truncatedLocale);
      if (properties == PROPERTIES_NOT_FOUND
              && getUserLocale().equals(locale)
              && truncatedLocale != null) {
          properties = loadPropertiesFromFile(packPath, locale, null);
      }
      ```
  * `AbstractMessages#searchClasspath`
    Replace `Properties properties = loadPropertiesFromFile(packPath, locale, truncatedLocale);` line with:
      ```
      Properties properties = loadPropertiesFromResource(packPath, locale, truncatedLocale);
      if (properties == PROPERTIES_NOT_FOUND
              && getUserLocale().equals(locale)
              && truncatedLocale != null) {
          properties = loadPropertiesFromResource(packPath, locale, null);
      }
      ```
Then, if the application does not find the properties (`@include` stores in properties) in the `messages_ru.properties` file for default locale (ru) , then the application will look for the properties in the `messages.properties` file.

Reply:
As far as I see this solution does not work if locale is specified explicitly in getMessage call. For example, `testIncludeDefaultLocale` fails if the following assertion is added:
```
            msg = messages.getMessage("com.haulmont.cuba.core.mp_test", "includedMsg", Locale.forLanguageTag("ru"));
            assertEquals("Included Message RU", msg);
```

Reply:
The case suggested above should not work. You have set the default English locale. The `@include=com.haulmont.cuba.core.mp_test.inc` is located in the `com\haulmont\cuba\core\mp_test\messages.properties` file. File `com\haulmont\cuba\core\mp_test\messages_ru.properties` for Russian locale does not exist. The platform first tries to find the `@include` in the `com\haulmont\cuba\core\mp_test\messages_ru.properties` file for the Russian locale, but since there is no file, and the Russian locale is not installed by default, the search for the message in Russian ends. Then the platform tries to find `@install` for the default locale (EN) and finds it. But since the default locale is English, then the search for the message goes in English.

To make your case work you need to either set the Russian locale as the default locale, as you did in the next test, or create the `com\haulmont\cuba\core\mp_test\messages_ru.properties` file and add an `@include` to it.

Reply:
I thought `@include` key should apply to all locales automatically. It is not convinient to specify the same includes in all language files.

I updated the fix and implemented extended search for default locale only. Also I changed the test to check the problem identical to described in the original issue.